File: ts/src/output/schemaValidator.ts
/**
 * ts/src/output/schemaValidator.ts
 *
 * Pure functions: stripMarkdownFences + validateWithSchema.
 * Used by queryFull() to coerce LLM text → validated structured output.
 *
 * Requirements: OUT-02 (validation) + OUT-03 (used inside retry loop).
 */

import { convertJsonSchemaToZod } from 'zod-from-json-schema';

/**
 * Regex for stripping ```json ... ``` or ``` ... ``` wrappers from LLM responses.
 * Handles optional 'json' language tag and CRLF line endings.
 */
const FENCE_RE = /^```(?:json)?\r?\n?([\s\S]*?)\r?\n?```$/s;

export function stripMarkdownFences(text: string): string {
  const trimmed = text.trim();
  const match = FENCE_RE.exec(trimmed);
  return match ? match[1].trim() : trimmed;
}

export type ValidationResult =
  | { success: true; data: unknown }
  | { success: false; error: string };

/**
 * Strip fences → JSON.parse → Zod validate against JSON Schema.
 *
 * Returns:
 *   { success: true, data } on valid JSON matching the schema.
 *   { success: false, error } on JSON parse failure OR schema mismatch.
 *     error is a non-empty string suitable for feeding to buildRetryPrompt.
 */
export function validateWithSchema(
  schema: Record<string, unknown>,
  text: string,
): ValidationResult {
  const stripped = stripMarkdownFences(text);
  let parsed: unknown;
  try {
    parsed = JSON.parse(stripped);
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return { success: false, error: `JSON parse failed: ${msg}` };
  }
  const zodSchema = convertJsonSchemaToZod(schema);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const result = (zodSchema as any).safeParse(parsed);
  if (result.success) {
    return { success: true, data: result.data };
  }
  // Zod v4 exposes .issues on the error; .message is the human-readable summary.
  const errMsg = result.error?.message ?? 'schema validation failed';
  return { success: false, error: errMsg };
}

File: python/src/gemini_sdk/output/schema_validator.py
"""python/src/gemini_sdk/output/schema_validator.py

Pure functions: strip_markdown_fences, validate_with_schema.
Mirrors ts/src/output/schemaValidator.ts.

Requirements: OUT-02, OUT-03 (used inside retry loop).
"""

from __future__ import annotations

import json
import re
from typing import Any, Tuple

import jsonschema

# Match ```json ... ``` or ``` ... ``` wrappers, with optional language tag
# and CRLF line endings. Mirrors TS FENCE_RE.
_FENCE_RE = re.compile(r"^```(?:json)?\r?\n?([\s\S]*?)\r?\n?```$", re.DOTALL)


def strip_markdown_fences(text: str) -> str:
    """Strip ```json ... ``` or ``` ... ``` wrappers from LLM output."""
    trimmed = text.strip()
    m = _FENCE_RE.match(trimmed)
    return m.group(1).strip() if m else trimmed


def validate_with_schema(schema: dict, text: str) -> Tuple[bool, Any, str]:
    """Strip fences -> json.loads -> jsonschema.validate.

    Returns:
      (True, data, "")          on valid JSON matching the schema.
      (False, None, <message>)  on JSON parse failure OR schema mismatch.
        <message> is a non-empty string suitable for feeding to build_retry_prompt.
    """
    stripped = strip_markdown_fences(text)
    try:
        parsed = json.loads(stripped)
    except json.JSONDecodeError as e:
        return False, None, f"JSON parse failed: {e}"

    try:
        jsonschema.validate(instance=parsed, schema=schema)
        return True, parsed, ""
    except jsonschema.ValidationError as e:
        return False, None, e.message
